Add explanation for Machine Config Pool Management (#52)
appuio/component-openshift4-nodes: docs/modules/ROOT/pages/explanations/machine-config-pool.adoc

= Managing Machine Config Pools

> This component allows you to manage machine config pool and all related configuration, including `MachineConfigs`, `KubeletConfigs`, and `ContainerRuntimeConfigs`.

Reviewer: ```suggestion
This component allows you to manage machine config pools and all related configuration, including `MachineConfigs`, `KubeletConfigs`, and `ContainerRuntimeConfigs`.
```

> If you wan to get a better understanding of machine config pools, please read the https://docs.openshift.com/container-platform/4.7/post_installation_configuration/machine-configuration-tasks.html[upstream documentation] or https://www.redhat.com/en/blog/openshift-container-platform-4-how-does-machine-config-pool-work[this blog post from RedHat].

Reviewer: Might make sense to link to `latest` here:

```suggestion
If you wan to get a better understanding of machine config pools, please read the https://docs.openshift.com/container-platform/latest/post_installation_configuration/machine-configuration-tasks.html[upstream documentation] or https://www.redhat.com/en/blog/openshift-container-platform-4-how-does-machine-config-pool-work[this blog post from RedHat].
```

On this page we'll provide some explanations and examples on how to configure exiting or add new machine config pools.

== Default Machine Config Pools

On a default APPUiO OpenShift 4 cluster there are five machine config pools.

There are two that are always present and managed by OpenShift directly.
These can't be removed or modified through the component.

* The *master* pool contains all master nodes.
* The *worker* pool contains all other nodes.

On an APPUiO OpenShift 4 cluster the *worker* pool is further split into different sub-pools

* The *app* pool contains the actual _worker_ nodes that are used to run the users applications.
* The *infra* pool contains the infra nodes that run OpenShift 4 workloads, such as Prometheus or Elasticsearch,
* The *storage* pool contains the Ceph cluster deployed by component rook ceph.
These nodes aren't always present, but the pool always exists.

[NOTE]
====
Being a sub-pool of the worker pool means it will inherit all configuration of the worker pool, but it will still count as a separate pool.
====

== Configuring Machine Config Pools

> The component allows you to configure machine config pools and their related resource through the `machineConfigPools` parameter.

Reviewer: ```suggestion
The component allows you to configure machine config pools and their related resources through the `machineConfigPools` parameter.
```

The component provides sane defaults for machine config and node selectors, which you generally shouldn't need to change.
+
NOTE: This can't be changed for the _master_ or _worker_ pool.

* There is one https://docs.openshift.com/container-platform/4.11/rest_api/machine_apis/kubeletconfig-machineconfiguration-openshift-io-v1.html[`KubeletConfig` resource] per machine config pool that can be managed through the `kubelet` parameter.
The parameter contains the complete spec of the `KubeletConfig`.
The `machineConfigPoolSelector` is set by the component and doesn't need to be modified.

* There is also one https://docs.openshift.com/container-platform/4.11/rest_api/machine_apis/containerruntimeconfig-machineconfiguration-openshift-io-v1.html[`ContainerRuntimeConfig` resource] per machine config pool that can be managed through the `containerRuntime` parameter.
The parameter contains the complete spec of the `ContainerRuntimeConfig`.
The `machineConfigPoolSelector` is again set by the component and doesn't need to be modified.

> * An arbitrary number of https://docs.openshift.com/container-platform/4.11/rest_api/machine_apis/machineconfig-machineconfiguration-openshift-io-v1.html[`MachineConfig` resources] can be managed through the `machineConfigs` parameter.

Reviewer: ```suggestion
* An arbitrary number of https://docs.openshift.com/container-platform/latest/rest_api/machine_apis/machineconfig-machineconfiguration-openshift-io-v1.html[`MachineConfig` resources] can be managed through the `machineConfigs` parameter.
```

The component handles selectors and labels so that these machine configs are assigned to the corresponding pool.

[IMPORTANT]
====
Changes that impact a nodes configuration, such as modifying kubelet config or container runtime config, will restart all nodes in the machine config pool.
====

Below you can see a few example configurations that should illustrate how you can manage existing machine config pools.

=== Example: Modifying App Pool

> Let's assume we're running a large cluster with more than 10 large app nodes.

Reviewer: ```suggestion
Let's assume that we're running a large cluster with more than 10 large app nodes.
```
